annotation-processor/src/main/java/org/hibernate/validator/ap/classchecks/AbstractMethodOverrideCheck.java

```diff
@@ -88,105 +79,88 @@ public AbstractMethodOverrideCheck(Elements elementUtils, Types typeUtils, Const
+	 * @param methodInheritanceTreeBuilder the method inheritance tree builder
 	 */
-	private Map<TypeElement, ExecutableElement> findOverriddenMethodInInterfacesPairs(
-			TypeElement currentTypeElement,
-			List<? extends TypeMirror> interfaces,
-			ExecutableElement currentMethod) {
-		Map<TypeElement,ExecutableElement> elements = CollectionHelper.newHashMap();
-
-		for ( TypeMirror anInterface : interfaces ) {
-			TypeElement implementedInterface = (TypeElement) typeUtils.asElement( anInterface );
-			ExecutableElement element = getOverriddenElement( currentTypeElement, implementedInterface, currentMethod );
-			if ( element != null ) {
-				elements.put( implementedInterface, element );
+	private void collectOverriddenMethodsInInterfaces(ExecutableElement overridingMethod, TypeElement currentTypeElement,
+			MethodInheritanceTree.Builder methodInheritanceTreeBuilder) {
+		for ( TypeMirror implementedInterface : currentTypeElement.getInterfaces() ) {
+			TypeElement interfaceTypeElement = (TypeElement) typeUtils.asElement( implementedInterface );
+			ExecutableElement overriddenMethod = getOverriddenMethod( overridingMethod, interfaceTypeElement );
+			if ( overriddenMethod != null ) {
+				methodInheritanceTreeBuilder.addOverriddenMethod( overridingMethod, overriddenMethod );
+				collectOverriddenMethodsInInterfaces( overriddenMethod, interfaceTypeElement, methodInheritanceTreeBuilder );
+			}
+			else {
+				collectOverriddenMethodsInInterfaces( overridingMethod, interfaceTypeElement, methodInheritanceTreeBuilder );
 			}
 		}
-
-		return elements;
 	}
 
 	/**
 	 * Find a method that is overridden by the one passed to this function.
 	 *
-	 * @param currentTypeElement a class in which method is located
-	 * @param otherTypeElement a class/interface on which to look for overridden method
 	 * @param currentMethod the method for which we want to find the overridden methods
-	 *
-	 * @return an overridden method if there's one, and {@code null} otherwise
+	 * @param typeElement the class or interface analyzed
+	 * @return the overridden method if there is one, and {@code null} otherwise
 	 */
-	private ExecutableElement getOverriddenElement(
-			TypeElement currentTypeElement,
-			TypeElement otherTypeElement,
-			ExecutableElement currentMethod) {
-
-		if ( isJavaLangObjectOrNull( otherTypeElement ) ) {
+	private ExecutableElement getOverriddenMethod(ExecutableElement currentMethod, TypeElement typeElement) {
+		if ( typeElement == null ) {
 			return null;
 		}
-
-		for ( Element element : elementUtils.getAllMembers( otherTypeElement ) ) {
+		for ( Element element : elementUtils.getAllMembers( typeElement ) ) {
 			if ( !element.getKind().equals( ElementKind.METHOD ) ) {
 				continue;
 			}
-			if ( elementUtils.overrides( currentMethod, (ExecutableElement) element, currentTypeElement ) ) {
+			if ( elementUtils.overrides( currentMethod, (ExecutableElement) element, getEnclosingTypeElement( currentMethod ) ) ) {
```

Comment:
maybe get enclosing element before the loop ? or this operation is not very expensive ? Something like:
```
TypeElement methodEnclosingType = getEnclosingTypeElement( currentMethod );
for (...){
....
if ( elementUtils.overrides( currentMethod, (ExecutableElement) element, methodEnclosingType ) ) {
```